adcc/LazyMp.py

```diff
@@ -60,39 +63,103 @@ def __getattr__(self, attr):
+        hf = self.reference_state
+        expressions = {
+            # space + contraction
+            b.ooov + b.vv: ('ijbc,kabc->ijka', b.ovvv),
+            b.ooov + b.ov: ('ilab,lkjb->ijka', b.ooov),
+            b.oovv + b.oo: ('klab,ijkl->ijab', b.oooo),
+            b.oovv + b.ov: ('jkac,kbic->ijab', b.ovov),
+            b.oovv + b.vv: ('ijcd,abcd->ijab', b.vvvv),
+            b.ovvv + b.oo: ('jkbc,jkia->iabc', b.ooov),
+            b.ovvv + b.ov: ('ijbd,jcad->iabc', b.ovvv),
+        }
+        contraction, eri_block = expressions[space + contraction]
+        return einsum(contraction, self.t2oo, hf.eri(eri_block))
 
-    @property
+    @cached_property
+    @timed_member_call(timer="timer")
     def mp2_diffdm(self):
         """
         Return the MP2 differensce density in the MO basis.
         """
-        ret = OneParticleOperator.from_cpp(self.__cpp.mp2_diffdm)
+        hf = self.reference_state
+        ret = OneParticleOperator(self.mospaces, is_symmetric=True)
+        ret[b.oo] = -0.5 * einsum("ikab,jkab->ij", self.t2oo, self.t2oo)
+        ret[b.ov] = -0.5 * (
+            + einsum("ijbc,jabc->ia", self.t2oo, hf.ovvv)
+            + einsum("jkib,jkab->ia", hf.ooov, self.t2oo)
+        ) / self.df(b.ov)
+        ret[b.vv] = 0.5 * einsum("ijac,ijbc->ab", self.t2oo, self.t2oo)
+
+        if self.has_core_occupied_space:
+            # additional terms to "revert" CVS for ground state density
+            ret[b.oo] += -0.5 * einsum("iLab,jLab->ij", self.t2oc, self.t2oc)
+            ret[b.ov] += -0.5 * (
+                + einsum("jMib,jMab->ia", hf.ocov, self.t2oc)
+                + einsum("iLbc,Labc->ia", self.t2oc, hf.cvvv)
+                + einsum("kLib,kLab->ia", hf.ocov, self.t2oc)
+                + einsum("iMLb,LMab->ia", hf.occv, self.t2cc)
+                - einsum("iLMb,LMab->ia", hf.occv, self.t2cc)
+            ) / self.df(b.ov)
+            ret[b.vv] += (+ 0.5 * einsum("IJac,IJbc->ab", self.t2cc, self.t2cc)
+                          + 1.0 * einsum("kJac,kJbc->ab", self.t2oc, self.t2oc))
+            # compute extra CVS blocks
+            ret[b.cc] = -0.5 * (+ einsum("kIab,kJab->IJ", self.t2oc, self.t2oc)
+                                + einsum('LIab,LJab->IJ', self.t2cc, self.t2cc))
+            ret[b.co] = -0.5 * (+ einsum("kIab,kjab->Ij", self.t2oc, self.t2oo)
+                                + einsum("ILab,jLab->Ij", self.t2cc, self.t2oc))
+            ret[b.cv] = -0.5 * (
+                - 1.0 * einsum("jIbc,jabc->Ia", self.t2oc, hf.ovvv)
+                + 1.0 * einsum("jkIb,jkab->Ia", hf.oocv, self.t2oo)
+                + 1.0 * einsum("jMIb,jMab->Ia", hf.occv, self.t2oc)
+                + 1.0 * einsum("ILbc,Labc->Ia", self.t2cc, hf.cvvv)
+                + 1.0 * einsum("kLIb,kLab->Ia", hf.occv, self.t2oc)
+                + 1.0 * einsum("LMIb,LMab->Ia", hf.cccv, self.t2cc)
+            ) / self.df(b.cv)
+        for bb in ret.blocks_nonzero:
+            ret[bb].evaluate()
```

Comment:
If we don't yet support ret.evaluate() directly, I would add it.

Reply:
Also I would set ret.reference_state and then just return evaluate(ret), but that's a matter of taste, just makes the flow clearer to me somehow.

Reply:
It's already there, so this is resolved 😄